orca/service_impl.go

```diff
@@ -0,0 +1,173 @@
+// corresponding to backend metrics, which eventually get pushed to clients who
+// have initiated the SteamCoreMetrics streaming RPC.
+//
+// [ORCA]: https://github.com/cncf/xds/blob/main/xds/service/orca/v3/orca.proto
+type Server struct {
+	v3orcaservicegrpc.UnimplementedOpenRcaServiceServer
+
+	minReportingInterval time.Duration
+	recorder             *metricRecorder
+}
+
+// ServerOptions contains options to configure the ORCA server implementation.
+type ServerOptions struct {
+	// MinReportingInterval sets the lower bound for how often out-of-band
+	// metrics are reported on the streaming RPC initiated by the client.  If
+	// unspecified or less than the default value of 30s, the default is used.
+	// Clients may request a higher value as part of the StreamCoreMetrics
+	// streaming RPC.
+	MinReportingInterval time.Duration
+}
+
+// Register creates a new ORCA service implementation configured using the
+// provided options and registers the same on the provided service registrar.
+func Register(s grpc.ServiceRegistrar, opts ServerOptions) (*Server, error) {
+	// TODO(easwars): Once the generated pb.gos are updated, we wont need this
+	// type assertion any more, since the new versions accept a
+	// grpc.ServiceRegistrar in their registration functions.
+	srv, ok := s.(*grpc.Server)
+	if !ok {
+		return nil, fmt.Errorf("concrete type of provided grpc.ServiceRegistrar is %T, only supported type is %T", s, &grpc.Server{})
+	}
+	server := NewServer(opts)
+	v3orcaservicegrpc.RegisterOpenRcaServiceServer(srv, server)
+	return server, nil
+}
+
+// NewServer returns a new ORCA service implementation configured using the
+// provided options. This can be used to customize the registration of the
+// returned ORCA server. Most usages should prefer to use Register instead.
+func NewServer(opts ServerOptions) *Server {
+	if opts.MinReportingInterval < internal.MinORCAReportingInterval {
+		opts.MinReportingInterval = internal.MinORCAReportingInterval
+	}
+	return &Server{
+		minReportingInterval: opts.MinReportingInterval,
+		recorder:             newMetricRecorder(),
+	}
+}
+
+func (s *Server) determineReportingInterval(req *v3orcaservicegrpc.OrcaLoadReportRequest) time.Duration {
+	if req.GetReportInterval() == nil {
+		return s.minReportingInterval
+	}
+	interval := req.GetReportInterval()
+	if err := interval.CheckValid(); err != nil {
+		logger.Warningf("Received reporting interval %q is invalid: %v. Using default: %s", interval, s.minReportingInterval)
+		return s.minReportingInterval
+	}
+	return interval.AsDuration()
+}
+
+func (s *Server) sendMetricsResponse(stream v3orcaservicegrpc.OpenRcaService_StreamCoreMetricsServer) error {
+	// The oobRecorder global singleton is guaranteed to be initialized at this
+	// point because registration of this service implementation and the
+	// initialization of the former take place in the same function,
+	// EnableOutOfBandMetricsReporting().
+	resp := s.recorder.toLoadReportProto()
```

Comment:
Optional: delete the local since it's all so trivial?

Reply:
Done.